src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/TreeMaker.java

```diff
@@ -80,23 +81,29 @@ public void freeze() {
         Declaration.Scoped make(Position pos, String name, Declaration... decls);
     }
 
-    interface VarFactoryNoLayout {
-        Declaration.Variable make(Position pos, String name, Type type);
+    Map<String, List<ConstantDesc>> collectAttributes(Cursor c) {
+        return c.children().filter(Cursor::isAttribute)
+                .collect(Collectors.groupingBy(
+                        attr -> attr.kind().name(),
+                        Collectors.mapping(Cursor::spelling, Collectors.toList())
+                ));
     }
 
     public Declaration createTree(Cursor c) {
-        switch (Objects.requireNonNull(c).kind()) {
+        Objects.requireNonNull(c);
+        Map<String, List<ConstantDesc>> attrs = collectAttributes(c);
```

Comment:
Rather than collecting attributes upfront, and then having to pass down the attributes to all the various createXYZ functions, wouldn't it be easier to take the parsed declaration and adding the attributes after the fact? That way the createXYZ could remain unchanged and the attribute changes would be more orthogonal. This could be implemented on top of either a public or an impl-only withAttribute functionality.

Reply:
That's certainly possible, but the change won't be that straightforward. We will need to change the checkCache mechanism as what need to be in cache should be with attributes after parsing.

Thus I don't think the change will be smaller, and the constructor with map are still needed. I'll give it a go and see how we feel about that.